cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java

```diff
@@ -1135,8 +1174,47 @@ private String getHiveDatabase(@Nullable String namespace) {
     if (namespace == null) {
       return null;
     }
-    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
-    return namespace.equals(Id.Namespace.DEFAULT.getId()) ? namespace : String.format("%s_%s", tablePrefix, namespace);
+    try {
+      if (namespace.equals(Id.Namespace.DEFAULT.getId())) {
+        return namespace;
+      } else {
+        String hiveDatabase = getCustomHiveDatabaseIfExists(namespace);
+        if (!Strings.isNullOrEmpty(hiveDatabase)) {
+          return hiveDatabase;
+        } else {
+          String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
+          return String.format("%s_%s", tablePrefix, namespace);
+        }
+      }
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Return the custom database for a given namespace if one exists else return null
+   *
+   * @param namespace the namespace for which we need to look up custom mapping
+   * @return the custom hive database name if one exists else null
+   * @throws ExploreException if failed to get NamespaceMeta for the given namespace
+   */
+  @Nullable
+  private String getCustomHiveDatabaseIfExists(@Nullable String namespace) throws ExploreException {
```

Comment:
This method wraps the logic of getting customNamespace if one exists. Without this we will need to handle  exception in create() and getHiveDatabase() because if we simply call getHiveDatabase() from create we don't know if it had custom mapping or not unless we check first which involves two calls to namespaequery admin. But  mostly this just wraps the getting meta part  to be called from different places. The return value seems logical to me if there is custom mapping return that if there is none return null to tell the caller that there isn't one. Now the parameter can be null because in ExploreService null namespace signifies  across all namespace we cannot do anything about that. Let me know if its sounds good or else we can think of something else.